File: Programmers/Programmers258711.cpp
// Programmers258711.cpp
#include <vector>

using namespace std;

bool visited[1'000'001] = { false, };
vector<int> adjList[1'000'001];

void dfs(int curr) {
    if (adjList[curr].size() != 0) {
        int next = adjList[curr][0];
        if (!visited[next]) {
            visited[next] = true;
            dfs(next);
        }
    }
}

vector<int> solution(vector<vector<int>> edges) {
    vector<int> ans(4);

    int n = 0;
    bool isNode[1'000'001] = { false, };
    int indegree[1'000'001] = { 0, }, outdegree[1'000'001] = { 0, };
    for (vector<int>& edge : edges) {
        n = max(n, max(edge[0], edge[1]));
        isNode[edge[0]] = true;
        isNode[edge[1]] = true;

        adjList[edge[0]].push_back(edge[1]);
        outdegree[edge[0]]++;
        indegree[edge[1]]++;
    }

    for (int i = 1; i <= n; i++) {
        if (indegree[i] == 0 && outdegree[i] >= 2) {
            ans[0] = i;
            for (int& adj : adjList[i])
                indegree[adj]--;
            break;
        }
    }

    vector<int> stick; // indegree가 0개이고 outdegree가 1개 이하인 노드들
    vector<int> graph8; // outdegree가 2개인 노드들
    for (int i = 1; i <= n; i++) {
        if (isNode[i] && indegree[i] == 0 && outdegree[i] <= 1)
            stick.push_back(i);
        else if (indegree[i] == 2 && outdegree[i] == 2)
            graph8.push_back(i);
    }

    for (int& i : stick) {
        visited[i] = true;
        dfs(i);
        ans[2]++;
    }

    for (int& i : graph8) {
        visited[i] = true;
        visited[adjList[i][0]] = true;
        visited[adjList[i][1]] = true;

        dfs(adjList[i][0]);
        dfs(adjList[i][1]);

        ans[3]++;
    }

    visited[ans[0]] = true;
    for (int i = 1; i <= n; i++) {
        if (isNode[i] && !visited[i]) {
            visited[i] = true;
            dfs(i);
            ans[1]++;
        }
    }

    return ans;
}

File: Programmers/Programmers258711.py
# Programmers258711.py
import sys
sys.setrecursionlimit(10 ** 6)

MAX_NODE = 1_000_001

visited = [False] * MAX_NODE
adj_list = [[] for _ in range(MAX_NODE)]

def dfs(curr):
    if adj_list[curr]:
        next_node = adj_list[curr][0]
        if not visited[next_node]:
            visited[next_node] = True
            dfs(next_node)

def solution(edges):
    ans = [0, 0, 0, 0]

    n = 0
    is_node = [False] * MAX_NODE
    indegree, outdegree = [0] * MAX_NODE, [0] * MAX_NODE
    for u, v in edges:
        n = max(n, u, v)

        is_node[u] = True
        is_node[v] = True

        adj_list[u].append(v)
        outdegree[u] += 1
        indegree[v] += 1

    for i in range(1, n + 1):
        if indegree[i] == 0 and outdegree[i] >= 2:
            ans[0] = i
            for adj in adj_list[i]:
                indegree[adj] -= 1
            break

    stick, graph8 = [], []
    for i in range(1, n + 1):
        if is_node[i] and indegree[i] == 0 and outdegree[i] <= 1:
            stick.append(i)
        elif indegree[i] == 2 and outdegree[i] == 2:
            graph8.append(i)

    for i in stick:
        visited[i] = True
        dfs(i)
        ans[2] += 1

    for i in graph8:
        visited[i] = True
        visited[adj_list[i][0]] = True
        visited[adj_list[i][1]] = True

        dfs(adj_list[i][0])
        dfs(adj_list[i][1])

        ans[3] += 1

    visited[ans[0]] = True
    for i in range(1, n + 1):
        if is_node[i] and not visited[i]:
            visited[i] = True
            dfs(i)
            ans[1] += 1

    return ans